common/splice_script.c

```diff
@@ -0,0 +1,2696 @@
+			/* Check for duplicate channel ids */
+			for (size_t j = 0; j < tal_count(chan_ids); j++)
+				if (input[i]->middle->type == TOK_CHANID
+					&& channel_id_eq(&chan_ids[j],
+							 input[i]->middle->chan_id))
+					return new_error(ctx, DUPLICATE_CHANID,
+							 input[i]->middle,
+							 "validate_and_clean");
+
+			/* Now add our channel id to the array */
+			if (input[i]->middle->type == TOK_CHANID)
+				tal_arr_expand(&chan_ids,
+					       *input[i]->middle->chan_id);
+			if (!is_valid_middle(input[i]->middle))
+				return new_error(ctx, INVALID_MIDDLE_OP,
+						 input[i]->middle,
+						 "validate_and_clean");
+			tokens[n++] = tal_steal(tokens, input[i]);
+			break;
+		}
+	}
+
+	tal_free(chan_ids);
+	tal_free(input);
+	tal_resize(&tokens, n);
+	*tokens_inout = tokens;
+	return NULL;
+}
+
+static struct splice_script_error *calculate_amounts(const tal_t *ctx,
+						     struct token ***tokens_inout)
+{
+	struct token **input = *tokens_inout;
+	struct token **tokens;
+	size_t n, left_wilds, right_wilds;
+	u32 left_used_ppm, right_used_ppm;
+	struct amount_sat left_total, right_total, usable_left;
+	struct token *left_wildcard_token, *wallet_token;
+
+	left_wilds = 0;
+	right_wilds = 0;
+	left_used_ppm = 0;
+	right_used_ppm = 0;
+	left_total = AMOUNT_SAT(0);
+	right_total = AMOUNT_SAT(0);
+	left_wildcard_token = NULL;
+	for (size_t i = 0; i < tal_count(input); i++) {
+		switch(input[i]->type) {
+		case TOK_CHAR:
+		case TOK_ATSYM:
+		case TOK_PLUS:
+		case TOK_MINUS:
+		case TOK_COLON:
+		case TOK_LEASERATE:
+		case TOK_ARROW:
+		case TOK_PIPE:
+		case TOK_STR:
+		case TOK_SATS:
+		case TOK_PERCENT:
+		case TOK_QUESTION:
+		case TOK_WILDCARD:
+		case TOK_FEE:
+		case TOK_CHANID:
+		case TOK_WALLET:
+		case TOK_FEERATE:
+		case TOK_NODEID:
+		case TOK_BTCADDR:
+		case TOK_CHANQUERY:
+		case TOK_MULTI_CHANID:
+		case TOK_LEASEREQ:
+		case TOK_DELIMITER:
+			return new_error(ctx, INVALID_TOKEN, input[i],
+					 "calculate_amounts");
+		case TOK_SEGMENT:
+			if (input[i]->left->type == TOK_WILDCARD) {
+				left_wildcard_token = input[i]->left;
+				left_wilds++;
+			}
+			if (input[i]->right->type == TOK_WILDCARD)
+				right_wilds++;
+			if (input[i]->left->type == TOK_SATS)
+				if (!amount_sat_add(&left_total, left_total,
+						    input[i]->left->amount_sat))
+					return new_error(ctx, INTERNAL_ERROR,
+							 input[i]->left,
+							 "calculate_amounts");
+			if (input[i]->right->type == TOK_SATS)
+				if (!amount_sat_add(&right_total, right_total,
+						    input[i]->right->amount_sat))
+					return new_error(ctx, INTERNAL_ERROR,
+							 input[i]->right,
+							 "calculate_amounts");
+			if (input[i]->left->type == TOK_PERCENT) {
+				left_used_ppm += input[i]->left->ppm;
+				if (left_used_ppm > 1000000)
+					return new_error(ctx,
+							 LEFT_PERCENT_OVER_100,
+							 input[i]->left,
+							 "calculate_amounts");
+			}
+			if (input[i]->right->type == TOK_PERCENT)
+				right_used_ppm += input[i]->right->ppm;
```

Comment:
nit: overflow check?

Reply:
`right_used_ppm` is only used to determine if there is any non-zero ppm amount so I believe this is fine.

Reply:
I'll switch it to a bool